### PR TITLE
added define config with OIDC_LOGIN_TYPE and OIDC_CLIENT_SCOPE

### PR DESCRIPTION
**Pull request by matchaxnb**
### All Submissions:

* [x] Have you followed the [plugin Contributing guideline](https://github.com/oidc-wp/openid-connect-generi/blob/dev/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

I add here support for define() settings for two variables: `login_type` and `scope`.

`OIDC_CLIENT_SCOPE` is a space-separated list of scopes.
`OIDC_LOGIN_TYPE` can either be `button` or `auto`. Other values have undefined behaviour.
`OIDC_ENFORCE_PRIVACY` (boolean): enable/disable the eponymous option
`OIDC_CREATE_IF_DOES_NOT_EXIST` (boolean) enable/disable the eponymous option
`OIDC_LINK_EXISTING_USERS` (boolean) enable/disable the eponymous option
`OIDC_REDIRECT_USER_BACK` (boolean) enable/disable the eponymous option
`OIDC_REDIRECT_ON_LOGOUT` (boolean) enable/disable the eponymous option

### How to test the changes in this Pull Request:

1. set `define('OIDC_CLIENT_SCOPE', 'email openid profile')` in your wp-config.php
2. see it shows up in the config and the variable is read-only
3. do the same for `OIDC_LOGIN_TYPE` and the other variables.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

Added seven new environment variables matching the options in the settings page.
Setting these `define`s disables changing the options in the interface.

`OIDC_CLIENT_SCOPE`: a space-separated list of scopes.
`OIDC_LOGIN_TYPE`: (`button`|`auto`)
`OIDC_ENFORCE_PRIVACY` (boolean): enable/disable the eponymous option
`OIDC_CREATE_IF_DOES_NOT_EXIST` (boolean) enable/disable the eponymous option
`OIDC_LINK_EXISTING_USERS` (boolean) enable/disable the eponymous option
`OIDC_REDIRECT_USER_BACK` (boolean) enable/disable the eponymous option
`OIDC_REDIRECT_ON_LOGOUT` (boolean) enable/disable the eponymous option

### CODE REVIEW

**Comment by ChloeTigre** [comment]
Hi there, the tests have failed but I have no idea why. It's a straight branch from develop and the changes are really minor.

**Comment by ChloeTigre** [comment]
opening a new issue.

**Comment by timnolte** [comment]
@ChloeTigre your PR is passing now. Will take a deeper look and consider it for inclusion in the next 3.9 feature release.

File: includes/openid-connect-generic-settings-page.php [hunk 1]
```diff
@@ -296,6 +298,7 @@ private function get_settings_fields() {
 				'title'       => __( 'Enforce Privacy', 'daggerhart-openid-connect-generic' ),
 				'description' => __( 'Require users be logged in to see the site.', 'daggerhart-openid-connect-generic' ),
 				'type'        => 'checkbox',
+				'disabled'    => defined( 'OIDC_ENFORCE_PRIVACY' ),
```
Comment (timnolte) [inline includes/openid-connect-generic-settings-page.php:301]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: includes/openid-connect-generic-settings-page.php [hunk 2]
```diff
@@ -347,24 +350,28 @@ private function get_settings_fields() {
 				'title'       => __( 'Link Existing Users', 'daggerhart-openid-connect-generic' ),
 				'description' => __( 'If a WordPress account already exists with the same identity as a newly-authenticated user over OpenID Connect, login as that user instead of generating an error.', 'daggerhart-openid-connect-generic' ),
 				'type'        => 'checkbox',
+				'disabled'    => defined( 'OIDC_LINK_EXISTING_USERS' ),
```
Comment (timnolte) [inline includes/openid-connect-generic-settings-page.php:353]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: includes/openid-connect-generic-settings-page.php [hunk 3]
```diff
@@ -347,24 +350,28 @@ private function get_settings_fields() {
...
 				'section'     => 'user_settings',
 			),
 			'create_if_does_not_exist'   => array(
 				'title'       => __( 'Create user if does not exist', 'daggerhart-openid-connect-generic' ),
 				'description' => __( 'If the user identity is not link to an existing Wordpress user, it is created. If this setting is not enabled and if the user authenticates with an account which is not link to an existing Wordpress user then the authentication failed', 'daggerhart-openid-connect-generic' ),
 				'type'        => 'checkbox',
+				'disabled'    => defined( 'OIDC_CREATE_IF_NOT_EXISTS' ),
```
Comment (timnolte) [inline includes/openid-connect-generic-settings-page.php:360]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: includes/openid-connect-generic-settings-page.php [hunk 4]
```diff
@@ -347,24 +350,28 @@ private function get_settings_fields() {
...
 				'section'     => 'user_settings',
 			),
 			'redirect_user_back'   => array(
 				'title'       => __( 'Redirect Back to Origin Page', 'daggerhart-openid-connect-generic' ),
 				'description' => __( 'After a successful OpenID Connect authentication, this will redirect the user back to the page on which they clicked the OpenID Connect login button. This will cause the login process to proceed in a traditional WordPress fashion. For example, users logging in through the default wp-login.php page would end up on the WordPress Dashboard and users logging in through the WooCommerce "My Account" page would end up on their account page.', 'daggerhart-openid-connect-generic' ),
 				'type'        => 'checkbox',
+				'disabled'    => defined( 'OIDC_REDIRECT_USER_BACK' ),
```
Comment (timnolte) [inline includes/openid-connect-generic-settings-page.php:367]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: includes/openid-connect-generic-settings-page.php [hunk 5]
```diff
@@ -347,24 +350,28 @@ private function get_settings_fields() {
...
 				'section'     => 'user_settings',
 			),
 			'redirect_on_logout'   => array(
 				'title'       => __( 'Redirect to the login screen when session is expired', 'daggerhart-openid-connect-generic' ),
 				'description' => __( 'When enabled, this will automatically redirect the user back to the WordPress login page if their access token has expired.', 'daggerhart-openid-connect-generic' ),
 				'type'        => 'checkbox',
+				'disabled'    => defined( 'OIDC_REDIRECT_ON_LOGOUT' ),
```
Comment (timnolte) [inline includes/openid-connect-generic-settings-page.php:374]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: openid-connect-generic.php [hunk 6]
```diff
@@ -344,13 +344,13 @@ static public function bootstrap() {
 				'identify_with_username' => false,
 
 				// Plugin settings.
-				'enforce_privacy' => 0,
+				'enforce_privacy' => defined( 'OIDC_ENFORCE_PRIVACY' ) ? intval( OIDC_ENFORCE_PRIVACY ) : 0,
```
Comment (timnolte) [inline openid-connect-generic.php:347]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: openid-connect-generic.php [hunk 7]
```diff
@@ -344,13 +344,13 @@ static public function bootstrap() {
...
 				'alternate_redirect_uri' => 0,
 				'token_refresh_enable' => 1,
-				'link_existing_users' => 0,
-				'create_if_does_not_exist' => 1,
-				'redirect_user_back' => 0,
-				'redirect_on_logout' => 1,
+				'link_existing_users' => defined( 'OIDC_LINK_EXISTING_USERS' ) ? intval( OIDC_LINK_EXISTING_USERS ) : 0,
+				'create_if_does_not_exist' => defined( 'OIDC_CREATE_IF_NOT_EXISTS' ) ? intval( OIDC_CREATE_IF_NOT_EXISTS ) : 1,
+				'redirect_user_back' => defined( 'OIDC_REDIRECT_USER_BACK' ) ? intval( OIDC_REDIRECT_USER_BACK ) : 0,
+				'redirect_on_logout' => defined( 'OIDC_REDIRECT_ON_LOGOUT' ) ? intval( OIDC_REDIRECT_ON_LOGOUT ) : 1,
```
Comment (timnolte) [inline openid-connect-generic.php:353]: @ChloeTigre you didn't document in your PR, and didn't add this to the PHPStan bootstap, adding this additional constant. Please remove this or add the missing documentation and additional support.

File: readme.md [hunk 8]
```diff
@@ -82,6 +82,8 @@ On the settings page for this plugin (Dashboard > Settings > OpenID Connect Gene
 - Userinfo Endpoint URL: `OIDC_ENDPOINT_USERINFO_URL`
 - Token Validation Endpoint URL: `OIDC_ENDPOINT_TOKEN_URL`
 - End Session Endpoint URL: `OIDC_ENDPOINT_LOGOUT_URL`
+- openid scope: `OIDC_CLIENT_SCOPE` (space separated)
+- openid login type: `OIDC_LOGIN_TYPE` ('button' or 'auto')
```
Comment (timnolte) [inline readme.md:86]: @ChloeTigre please ensure that these labels match those in the plugin settings screen.

**Review by timnolte — CHANGES_REQUESTED** [review]
@ChloeTigre comments made some changes are required. Thanks!

**Comment by ChloeTigre** [comment]
Yes doing that!

**Comment by ChloeTigre** [comment]
There :) thanks.

**Comment by ChloeTigre** [comment]
Hey there ! I have no time to care for the README.md change and I am unsure of the desired styling. Can someone take care of this? I'd like to get this merged but it's out of my professional scope at the moment.

**Review by timnolte — APPROVED** [review]

**Pull request merged**